File: ls/server/src/main/java/org/opencds/cqf/cql/ls/server/provider/FormattingProvider.java
package org.opencds.cqf.cql.ls.server.provider;

import java.io.IOException;
import java.net.URI;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import org.cqframework.cql.tools.formatter.CqlFormatterVisitor;
import org.cqframework.cql.tools.formatter.CqlFormatterVisitor.FormatResult;
import org.eclipse.lsp4j.Position;
import org.eclipse.lsp4j.Range;
import org.eclipse.lsp4j.TextEdit;
import org.opencds.cqf.cql.ls.core.ContentService;
import org.opencds.cqf.cql.ls.core.utility.Uris;

public class FormattingProvider {
    private ContentService contentService;

    public FormattingProvider(ContentService contentService) {
        this.contentService = contentService;
    }

    public List<TextEdit> format(String uri) {
        URI u = Objects.requireNonNull(Uris.parseOrNull(uri));

        FormatResult fr = null;
        try {
            fr = CqlFormatterVisitor.getFormattedOutput(this.contentService.read(u));
        } catch (IOException e) {
            throw new IllegalArgumentException("Unable to format CQL due to an error.", e);
        }

        if (!fr.getErrors().isEmpty()) {
            throw new IllegalArgumentException(
                    String.join("\n", "Unable to format CQL due to syntax errors.",
                            "Please fix the errors and try again."));
        }

        TextEdit te = new TextEdit(
                new Range(new Position(0, 0), new Position(Integer.MAX_VALUE, Integer.MAX_VALUE)),
                fr.getOutput());

        return Collections.singletonList(te);
    }
}

File: ls/server/src/main/kotlin/org/opencds/cqf/cql/ls/server/provider/FormattingProvider.kt
package org.opencds.cqf.cql.ls.server.provider

import org.cqframework.cql.tools.formatter.CqlFormatterVisitor
import org.eclipse.lsp4j.Position
import org.eclipse.lsp4j.Range
import org.eclipse.lsp4j.TextEdit
import org.opencds.cqf.cql.ls.core.ContentService
import org.opencds.cqf.cql.ls.core.utility.Uris

class FormattingProvider(private val contentService: ContentService) {
    fun format(uri: String): List<TextEdit> {
        val u = requireNotNull(Uris.parseOrNull(uri))

        val fr =
            try {
                CqlFormatterVisitor.getFormattedOutput(
                    contentService.read(u) ?: throw IllegalArgumentException("Unable to read content from: $u"),
                )
            } catch (e: Exception) {
                throw IllegalArgumentException("Unable to format CQL due to an error.", e)
            }

        if (fr.errors.isNotEmpty()) {
            throw IllegalArgumentException(
                listOf("Unable to format CQL due to syntax errors.", "Please fix the errors and try again.").joinToString("\n"),
            )
        }

        val te =
            TextEdit(
                Range(Position(0, 0), Position(Int.MAX_VALUE, Int.MAX_VALUE)),
                fr.output,
            )

        return listOf(te)
    }
}
